mlir/include/mlir/Dialect/Utils/IndexingUtils.h

```diff
@@ -202,6 +202,9 @@ SmallVector<T> applyPermutation(ArrayRef<T> input,
                                 ArrayRef<int64_t> permutation) {
   assert(input.size() == permutation.size() &&
          "expected input rank to equal permutation rank");
+  assert(
+      llvm::all_of(permutation, [&](size_t s) { return s < input.size(); }) &&
+      "permutation must be within input bounds");
```

Comment:
Top level question: Do we really need to do this check? The permutation itself represents its value range. If there's an error, it should have been caught before the call site. In what scenarios would we need this check?

Reply:
I feel this check is needed. If you look at the original issue ([https://github.com/llvm/llvm-project/issues/99513]) there was a crash with cryptic assertion fail. This is happening because of the way permutation range is constructed :
`  auto permutationRange = llvm::map_range(
      llvm::seq<unsigned>(0, input.size()),
      [&](int64_t idx) -> T { return input[permutation[idx]]; });`

We want to make sure that we are not accessing out of boud while doing "input[permutation[idx]]"